cidc_api/shared/auth.py

```diff
@@ -0,0 +1,322 @@
+        endpoint.is_protected = True
+
+        @wraps(endpoint)
+        def wrapped(*args, **kwargs):
+            is_authorized = check_auth(allowed_roles, resource, request.method)
+            if not is_authorized:
+                raise Unauthorized("Please provide proper credentials")
+            return endpoint(*args, **kwargs)
+
+        return wrapped
+
+    return decorator
+
+
+def public(endpoint):
+    """Declare an endpoint to be public, i.e., not requiring auth."""
+    # Store metadata on this function stating that it is unprotected
+    endpoint.is_protected = False
+
+    return endpoint
+
+
+def check_auth(allowed_roles: List[str], resource: str, method: str) -> bool:
+    """
+    Perform authentication and authorization for the current request. 
+    
+    Args:
+        allowed_roles: a list of CIDC user roles allowed to access this endpoint
+        resource: the resource targeted by this request
+        method: the HTTP method of this request
+    Returns:
+        bool, `True` if authentication and authorization passed.
+    """
+    user = authenticate()
+
+    try:
+        is_authorized = authorize(user, allowed_roles, resource, method)
+    except Unauthorized:
+        _log_user_and_request_details(False)
+        raise
+
+    _log_user_and_request_details(is_authorized)
+
+    _enforce_cli_version()
+
+    return is_authorized
+
+
+### Current user management ###
+CURRENT_USER_KEY = "current_user"
+
+
+def _set_current_user(user: Users):
+    """Store a user in the current request's context."""
+    assert isinstance(user, Users), "`user` must be an instance of the `Users` model"
+    setattr(g, CURRENT_USER_KEY, user)
+
+
+def get_current_user() -> Users:
+    """Returns the authenticated user who made the current request."""
+    current_user = g.get(CURRENT_USER_KEY)
+
+    assert current_user, (
+        "There is no user associated with the current request.\n"
+        "Note: `auth.get_current_user` can't be called by a request handler without authentication. "
+        "Decorate your handler with `auth.requires_auth` to authenticate the requesting user before calling the handler."
+    )
+
+    return current_user
+
+
+### Authentication logic ###
+_user_schema = UserSchema()
+
+
+def authenticate() -> Users:
+    id_token = _extract_token()
+    public_key = _get_issuer_public_key(id_token)
+    token_payload = _decode_id_token(id_token, public_key)
+    profile = {"email": token_payload["email"]}
+    return _user_schema.load(profile)
+
+
+def _extract_token() -> str:
+    """Extract an identity token from the current request's authorization header."""
+    auth_header = request.headers.get("Authorization")
+
+    if not auth_header or not auth_header.lower().startswith("bearer"):
+        raise Unauthorized(
+            "Authorization header must be set with structure 'Authorization: Bearer <id token>'"
+        )
+
+    id_token = auth_header.split(" ")[1]
```

Comment:
totally nitpick: good practice would be, I think, return "400 bad request" in case of malformed auth header ((or `unauth`, but not `500 server error`), which this will do for something like `Authorization: Bearer` without following token